AP_Proximity: add support for RP Lidar A2 (360° Lidar with upto 16m range)

amilcarlucas opened the pull request.

OXINARF commented:
@amilcarlucas Is this ready for review? It's marked as WIP...

amilcarlucas commented:
We are still going to do small changes to it, so please wait a while longer before reviewing.

khancyr commented:
@amilcarlucas ping !

amilcarlucas commented:
This is ready for review  since 11 July.
Is there anything I can do ?

khancyr commented:
it doesn't pass travis, can you look at it ?

amilcarlucas commented:
@stevenjosefs Can you please take a look at it ?

amilcarlucas commented:
This fixed the tests on our local CI server. Let's see how Travis-CI likes it :)

OXINARF commented:
@amilcarlucas 

> This is ready for review since 11 July.

You didn't write anything, we can't guess 🙂 

> This fixed the tests on our local CI server. Let's see how Travis-CI likes it :)

Well, it doesn't even build... You need to rebase on master and fix the issues.

OXINARF commented on libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:
```diff
@@ -0,0 +1,486 @@
+                if (convert_angle_to_sector(angle_deg, sector)) {
+                    _angle[sector] = angle_deg;
+                    _distance[sector] = distance_m;
+                    _distance_valid[sector] = true;
```
There's no check for distance being 0 and so invalid.

OXINARF commented on libraries/AP_Proximity/AP_Proximity_RPLidarA2.h:
```diff
@@ -0,0 +1,160 @@
+
+    // reply related variables
+    AP_HAL::UARTDriver *_uart;
+  //  char element_buf[2][10];
```
Don't leave dead code.

amilcarlucas commented on libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:
```diff
@@ -0,0 +1,489 @@
+extern const AP_HAL::HAL& hal;
+
+/*
+   The constructor also initialises the proximity sensor. Note that this
```
The SF40 codes does the same, the comment must be fixed in there as well

amilcarlucas commented on libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:
```diff
@@ -0,0 +1,489 @@
+// get maximum distance (in meters) of sensor
+float AP_Proximity_RPLidarA2::distance_max() const
+{
+    return 16.0f;  //16m max range RPLIDAR2
```
We only support the 16m version. That version is a special order, you have to explicitly say that you want that version, and then they sell it to you. No extra charge, the price is the same.

amilcarlucas commented on libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:
```diff
@@ -0,0 +1,486 @@
+    _uart->write(0x32);
+    _uart->write(0x32); //SP as padding for ftdi cable
+    _uart->write(0x32);
+#endif
```
When debuging with an FTDI cable attached. removed with a separated comit, so that it can be reverted if needed

amilcarlucas commented on libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:
```diff
@@ -0,0 +1,486 @@
+    uint8_t ignore_area_count = get_ignore_area_count();
+    if (ignore_area_count == 0) {
+        _sector_initialised = true;
+        return;
```
Again that is the same code as AP_Proximity_LightWareSF40C.cpp
That again needs to be fixed in there as well.

rmackay9 replied:
we need the special check so that if the "ignore_area_count" is zero, we leave the _num_sectors and _boundary array at their defaults which is 8 sectors of 45degrees each.  Without the special check the _num_sectors would be set to zero.

amilcarlucas commented on libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:
```diff
@@ -0,0 +1,486 @@
+}
+
+// initialise sector angles using user defined ignore areas, left same as SF40C
+// TODO: adapt to RPLIDAR A2 measurements
```
This function is a 1:1 copy of AP_Proximity_LightWareSF40C::init_sectors().

rmackay9 replied:
I think we can remove this To-Do.  The To-Do itself isn't in the SF40C version.

amilcarlucas commented:
We improved it a bit more and will push more comits soon.

amilcarlucas commented:
@OXINARF I think we have addressed all of your points. The missing ones need to be addressed by @rmackay9 because he is the author of the original SF40 code and the issues you raise are SF40 issues as well.

amilcarlucas commented on libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:
```diff
@@ -0,0 +1,452 @@
+    _uart->write(tx_buffer, 2);
+    _resetted = true;                   ///< be aware of extra 63 bytes coming after reset containing FW information
+    Debug(1, "LIDAR reset");
+    hal.scheduler->delay(RESET_RPA2_WAIT_MS);                      ///< wait at least 8ms after sending new request
```
Any Ideas on How to improve and or avoid this hal.scheduler->delay(RESET_RPA2_WAIT_MS); call ?

rmackay9 commented on libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:
```diff
@@ -0,0 +1,486 @@
+#if RP_DEBUG_LEVEL > 0
+    GCS_MAVLINK::send_text(MAV_SEVERITY_INFO, "LIDAR reset");
+#endif
+    hal.scheduler->delay(RESET_RPA2_WAIT_MS);                      ///< wait at least 8ms after sending new request
```
yes, this delay is a killer.  We definitely can't have this.  Instead we should record the time of the reset and don't try and process any messages until 8ms after that time.

rmackay9 commented:
I've made some formatting changes, squashed the commits and removed the killer delay (and added a To-Do in it's place) and pushed into master.  Thanks!
